relayer/message_relayer.go

```diff
@@ -548,27 +586,176 @@ func (r *messageRelayer) aggregateSignatures(signatureMap map[int]blsSignatureBu
+			zap.Uint64("latestProcessedBlock", latestProcessedBlock),
+		)
+		return latestProcessedBlock, nil
+	}
+	// Otherwise, return the configured start block height
+	r.logger.Info(
+		"Processing historical blocks from the configured start block height",
+		zap.String("relayerKey", r.relayerKey.GetKey().String()),
+		zap.Uint64("processHistoricalBlocksFromHeight", processHistoricalBlocksFromHeight),
+	)
+	return processHistoricalBlocksFromHeight, nil
+}
+
+// Gets the height of the chain head, writes it to the database, then returns it.
+func (r *messageRelayer) setProcessedBlockHeightToLatest() (uint64, error) {
+	ethClient, err := ethclient.Dial(r.sourceBlockchain.RPCEndpoint)
+	if err != nil {
+		r.logger.Error(
+			"Failed to dial node",
+			zap.String("blockchainID", r.sourceBlockchain.GetBlockchainID().String()),
+			zap.Error(err),
+		)
+		return 0, err
+	}
+
+	latestBlock, err := ethClient.BlockNumber(context.Background())
+	if err != nil {
+		r.logger.Error(
+			"Failed to get latest block",
+			zap.String("blockchainID", r.sourceBlockchain.GetBlockchainID().String()),
+			zap.Error(err),
+		)
+		return 0, err
+	}
+
+	r.logger.Info(
+		"Updating latest processed block in database",
+		zap.String("relayerKey", r.relayerKey.GetKey().String()),
+		zap.Uint64("latestBlock", latestBlock),
+	)
+
+	err = r.storeBlockHeight(latestBlock)
+	if err != nil {
+		r.logger.Error(
+			fmt.Sprintf("failed to put %s into database", database.LatestProcessedBlockKey),
+			zap.String("relayerKey", r.relayerKey.GetKey().String()),
+			zap.Error(err),
+		)
+		return 0, err
+	}
+	return latestBlock, nil
+}
+
+// Get the latest processed block height from the database.
+// Note that there may be unrelayed messages in the latest processed block
+// because it is updated as soon as a single message from that block is relayed,
+// and there may be multiple message in the same block.
+func (r *messageRelayer) getLatestProcessedBlockHeight() (uint64, error) {
+	latestProcessedBlockData, err := r.db.Get(r.relayerKey.GetKey(), database.LatestProcessedBlockKey)
+	if err != nil {
+		return 0, err
+	}
+	latestProcessedBlock, err := strconv.ParseUint(string(latestProcessedBlockData), 10, 64)
+	if err != nil {
+		return 0, err
+	}
+	return latestProcessedBlock, nil
+}
+
+// Store the block height in the database. Does not check against the current latest processed block height.
+func (r *messageRelayer) storeBlockHeight(height uint64) error {
+	return r.db.Put(r.relayerKey.GetKey(), database.LatestProcessedBlockKey, []byte(strconv.FormatUint(height, 10)))
+}
+
+// Stores the block height in the database if it is greater than the current latest processed block height.
+func (r *messageRelayer) storeLatestBlockHeight(height uint64) error {
+	// First, check that the stored height is less than the current block height
+	// This is necessary because the relayer may be processing blocks out of order on startup
+	latestProcessedBlock, err := r.getLatestProcessedBlockHeight()
```

Comment:
It will be nice to hopefully remove this extra read (even if just from memory) once we have #234 and #31